Source/Diagnostics/ComputeDiagFunctors/BackTransformParticleFunctor.cpp

```diff
@@ -0,0 +1,176 @@
+
+        for (WarpXParIter pti(*m_pc_src, lev); pti.isValid(); ++pti) {
+            auto ptile_dst = pc_dst.DefineAndReturnParticleTile(lev, pti.index(), pti.LocalTileIndex() );
+        }
+
+        auto& particles = m_pc_src->GetParticles(lev);
+#ifdef AMREX_USE_OMP
+#pragma omp parallel
+#endif
+        {
+            // Temporary arrays to store copy_flag and copy_index for particles
+            // that cross the z-slice
+            amrex::Gpu::DeviceVector<int> FlagForPartCopy;
+            amrex::Gpu::DeviceVector<int> IndexForPartCopy;
+
+            for (WarpXParIter pti(*m_pc_src, lev); pti.isValid(); ++pti) {
+
+                auto index = std::make_pair(pti.index(), pti.LocalTileIndex());
+
+                const auto GetParticleFilter = SelectParticles(pti, tmp_particle_data,
+                                               m_current_z_boost[i_buffer],
+                                               m_old_z_boost[i_buffer]);
+                const auto GetParticleLorentzTransform = LorentzTransformParticles(
+                                                         pti, tmp_particle_data,
+                                                         t_boost, dt,
+                                                         m_t_lab[i_buffer]);
+
+                long const np = pti.numParticles();
+
+                FlagForPartCopy.resize(np);
+                IndexForPartCopy.resize(np);
+
+                int* const AMREX_RESTRICT Flag = FlagForPartCopy.dataPtr();
+                int* const AMREX_RESTRICT IndexLocation = IndexForPartCopy.dataPtr();
+
+                const auto& ptile_src = particles.at(index);
+                auto src_data = ptile_src.getConstParticleTileData();
+                // Flag particles that need to be copied if they cross the z-slice
+                // setting this to 1 for testing (temporarily)
+                amrex::ParallelFor(np,
+                [=] AMREX_GPU_DEVICE(int i)
+                {
+                    Flag[i] = GetParticleFilter(src_data, i);
+                });
+
+                const int total_partdiag_size = amrex::Scan::ExclusiveSum(np,Flag,IndexLocation);
+                auto& ptile_dst = pc_dst.DefineAndReturnParticleTile(lev, pti.index(), pti.LocalTileIndex() );
+                auto old_size = ptile_dst.numParticles();
+                ptile_dst.resize(old_size + total_partdiag_size);
+                auto count = amrex::filterParticles(ptile_dst, ptile_src, GetParticleFilter, 0, old_size, np);
+                auto dst_data = ptile_dst.getParticleTileData();
+                amrex::ParallelFor(np,
+                [=] AMREX_GPU_DEVICE(int i)
+                {
+                   if (Flag[i] == 1) GetParticleLorentzTransform(dst_data, src_data, i,
+                                                                 old_size + IndexLocation[i]);
+                });
+                total_particles_added += count;
+            }
+        }
+        amrex::Gpu::synchronize();
```

Comment:
Wouldn't the destroyed `WarpXParIter` synchronize already? (Does it rly?)
Or is this handling a special case (like no particles)?

Reply:
I believe you are correct and this sync could be removed. Even if there are no particles, the `WarpXParIter` destructor will be called and that will call `amrex::Gpu::synchronize()`.